keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,960 @@
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. The KEP
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+#### Removing a Deprecated Flag
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+
+**For non-optional features moving to GA, the graduation criteria must include 
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+### Upgrade / Downgrade Strategy
```

Comment:
do we have any gotchas when upgrading a user from root CP to a non-root CP when doing kubeadm upgrade?

Reply:
+1 also have this question, it'd be great if we can think about that scenario and the scenario where you have mixed machines running as root and some running as non-root (I don't foresee any issues for that, but would like to make sure it's written down).

Reply:
For all user cases, kubeadm will need to check that a user/group is on the system, check the current filesystem, potentially reset ownership of files and deploy the new static pods.

Reply:
I was thinking we should just limit the scope of the feature to cluster create in alpha? So only for newly created cluster will control-plane as non-root will apply. On cluster upgrades and rollbacks we don't try to run the control-plane as non-root. Maybe we add that when this feature goes to beta?

Reply:
by tradition we always plan upgrade impacting features and enable them in alpha so that they can be tested ASAP.

we must document two cases.

**in-place upgrade:** where `kubeadm upgrade` is called on the same nodes to upgrade manifests

the way kubeadm upgrade works is it regenerates manifests based on settings in the current latest binary / kube-system:kubeadm-config:ClusterConfiguration 
it's not mutation of existing manifests on disk. so it might a bit more difficult to **not migrate** the control-plane manifests during upgrade.

given the feature gate will be alpha in e.g. 1.22, this means that e.g. a 1.21 control-plane will not be migrated to non-root unless the user patches kube-system:kubeadm-config:ClusterConfiguration feature-gate before doing the `kubeadm upgrade`. we need to figure out what are the gotchas with that.

@randomvariable is right. we just need to do the same checks for uid/gid existence during upgrade as when we are doing kubeadm init.

**re-place upgrade:** where new version nodes join the cluster to replace the old ones (ala Cluster API upgrades)

@vincepri 's comment is valid. we need to ensure the reader of this KEP that there will be no issues when having mixed root-non / root CP services during the node re-place upgrade process.